service/controller/resource/nodecontroller/create.go

```diff
@@ -21,107 +25,84 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+			WorkloadK8sClient:   k8sClient,
+			Logger:              r.logger,
+			Name:                fmt.Sprintf("%s-%s-nodes", project.Name(), clusterID),
+			Selector: labels.SelectorFromSet(map[string]string{
+				// When managing endpoints, we only consider node- and pod-readiness for workers. Master endpoint IPs
+				// are always present in the master endpoints object. This may change when we implement HA for KVM.
+				"role":                key.WorkerID,
+				label.OperatorVersion: project.Version(),
+			}),
+		}
+		desiredController, err = nodecontroller.New(config)
+		if err != nil {
+			return microerror.Mask(err)
+		}
 	}
 
-	return nil
-}
-
-// applyCreateChangeAsync boots the desired controller and saves it in the controllers map
-// if it boots successfully. It should be run as a goroutine to avoid blocking reconciliation of
-// other resources.
-func (r *Resource) applyCreateChangeAsync(ctx context.Context, desired controllerWithConfig) {
-	go desired.Boot(ctx)
-	<-desired.Booted() // Wait for the controller to be booted
-
 	r.controllerMutex.Lock()
-	controllerKey := controllerMapKey(desired.cluster)
-	if _, ok := r.controllers[controllerKey]; ok {
-		// There's a small chance for a race if two goroutines are booting controllers for the same clusters
-		// because, for example, several changes were made to a cluster CR in a short amount of time and booting
-		// of controllers takes a non-zero amount of time. By stopping the desired controller within the mutex,
-		// we ensure that only one controller is running at a time and we don't have dangling controllers causing
-		// a memory leak.
-		desired.Stop(ctx)
-	} else {
-		r.controllers[controllerKey] = desired
-	}
-	r.controllerMutex.Unlock()
-
-	r.logger.Debugf(ctx, "controller booted and registered")
-}
-
-// applyUpdateChange checks for an existing controller, stops it if it doesn't match the desired controller,
-// and returns a value indicating if the desired controller should be started (because the existing one was
-// stopped or there was no existing one).
-func (r *Resource) applyUpdateChange(ctx context.Context, desired controllerWithConfig) bool {
-	r.controllerMutex.Lock()         // Exclude access from other goroutines
-	defer r.controllerMutex.Unlock() // Ensure this is run for all function returns
-
-	controllerKey := controllerMapKey(desired.cluster)
-	current, ok := r.controllers[controllerKey]
-
-	if !ok {
-		r.logger.Debugf(ctx, "existing controller not found")
-		return true
+	defer r.controllerMutex.Unlock()
+
+	if current, ok := r.controllers[controllerKey]; ok {
+		var shouldStop bool
+		if !current.Equal(desiredController) {
+			r.logger.Debugf(ctx, "controllers don't match")
+			shouldStop = true
+		} else if time.Since(current.LastReconciled()) > 2*nodecontroller.ResyncPeriod {
+			r.logger.Debugf(ctx, "controller hasn't reconciled in more than %d minutes", nodecontroller.ResyncPeriod.Minutes())
+			shouldStop = true
+		}
+
+		if shouldStop {
```

Comment:
Is there a reason we're doing the `shouldStop` checks and actual stopping in two different places in the same function? Can we merge them?

Reply:
The code is not pretty either way. I got rid of the duplicate code in the latest commit at at least. Let me know what you think.

Reply:
The mutex around the map makes it difficult.